### ISSUE TITLE
COMPLETE_TRANSITION fired multiple times

### ISSUE THREAD

**Issue by jjdp** [react-navigation/react-navigation#3730]
So my navigator has a structure like this:

StackNavigator -> DrawerNavigator -> Each screen of the Drawer has a StackNavigator with multiple screens

I have noticed this on the new 1.5.2 version (upgraded from 1.0.0-beta.22). I'm guessing each navigation fires a COMPLETE_TRANSITION event since when DrawerToggle and DrawerOpen fires there are two COMPLETE_TRANSITION in the end.

But I have noticed when I navigate on the second stack the COMPLETE_TRANSITION fires twice, three on times on the third stack, so on and so forth. With the key of the root stack, in my case 'Profile' as shown in the image, and the 'StackRouterRoot'

Is this normal behavior?

### Current Behavior

- The COMPLETE_TRANSITION action fires multiple times

![a](https://user-images.githubusercontent.com/5244520/37329248-f7f07a1e-26d7-11e8-84fa-9aae46db7735.png)

### Expected Behavior

- Should fire once?

### Your Environment

Android 6.0 Emulator

| software         | version
| ---------------- | -------
| react-navigation | 1.5.2
| react-native     | 0.53.3
| node             | 6.11.2
| npm or yarn      | 1.5.1

**Comment by sraka1** [comment]
I'm getting the same behaviour. It looks like the events bubble up the tree and it looks like it's like that intentionally.

I.e. you have 10 StackNavigators layered on top of each other...you will get 10 COMPLETE_TRANSITION events, (each dispatched with the key of the navigator). This can slow down the app considerably by spamming the state with events...

I think there should be an option to switch off this behaviour entirely, or to only trigger it just with the key of the stack where the transition actually happened (and not with the parent navigators).

**Comment by brentvatne** [comment]
@ericvicenti ping

**Comment by brentvatne** [comment]
@sraka1 - I haven't investigated this but I don't suspect this slows the app down, however logging does slow it down so you may want to filter out the complete transition logs

**Comment by sraka1** [comment]
Exactly, the slowdown remark was referring to redux... 

As for your remark regarding redux, I can't imagine using react-navigation without it and with good reason. Discussion relevant for this issue though.

**Comment by brentvatne** [comment]
@sraka1 - can you why you feel you need redux in https://github.com/react-navigation/rfcs/issues/14 please?

**Comment by walker-ll** [comment]
same issue here，StackNavigator -> StackNavigator，in redux，dispatch navigate in a action，many times...and，end with:
```
Possible Unhandled Promise Rejection
Invariant Violation: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops...
```
how can i fix it?

**Comment by walker-ll** [comment]
I found this issue the same as mine. use redux navigator reducer、 immutableJS...
i fixed it with is(oldNavigatorState， newNavigatorState) from immutableJS method in navigator reducer， but its root cause remains unresolved.

**Comment by axelnormand** [comment]
same here. 
thanks @walker-ll I'll try that fix.

I personally use redux to so i can use redux-persist, fire actions from sagas etc.
Also IMHO conceptually routing is just state anyway so should be in the store :)

**Comment by lxcid** [comment]
I just found out about this issue it cause the unmounting and mounting my screen until it return this error out.

![screenshot 2018-03-30 12 18 27](https://user-images.githubusercontent.com/297156/38124488-80883fc6-3414-11e8-9b97-2e1446596cca.png)

Might be remotely related to #3867 as well.

**Comment by ericvicenti** [comment]
Good news, I've found a fix that should address most of the problems you're seeing:
https://github.com/react-navigation/react-navigation/pull/3902

To try the fix, please temporarily use this repo's master branch, or use the not-yet-released `2.0.0-beta.7` version

### To anybody using redux!

Please do not post bugs like this without a clear log of what actions and state have been rendered, and a log of all actions that come out of the view.

We cannot help you debug behavior in redux apps unless you show us exactly what state and actions are going back and forth between redux and the navigators!
